vsintegration/src/FSharp.Editor/LanguageService/LanguageService.fs

```diff
@@ -211,142 +242,194 @@ type
     inherit AbstractLanguageService<FSharpPackage, FSharpLanguageService>(package)
 
     let projectInfoManager = package.ComponentModel.DefaultExportProvider.GetExport<ProjectInfoManager>().Value
+    let solutionDTE = package.DTE.Solution
+
+    let mutable projectTracker = None : VisualStudioProjectTracker option
 
     let projectDisplayNameOf projectFileName = 
         if String.IsNullOrWhiteSpace projectFileName then projectFileName
         else Path.GetFileNameWithoutExtension projectFileName
+    
+    let singleFileProjects = ConcurrentDictionary<ProjectId, AbstractProject>()
 
-    let singleFileProjects = ConcurrentDictionary<_, AbstractProject>()
 
     let tryRemoveSingleFileProject projectId =
         match singleFileProjects.TryRemove(projectId) with
         | true, project ->
             projectInfoManager.RemoveSingleFileProject(projectId)
             project.Disconnect()
         | _ -> ()
+    
+    let getProject (projectId:ProjectId) =
+        self.Workspace.CurrentSolution.GetProject projectId
+
+
+    member self.OnWorkspaceChanged (args:WorkspaceChangeEventArgs) =
+        match args.Kind with
+        | WorkspaceChangeKind.SolutionAdded ->
+            args.NewSolution.Projects ?> Seq.iter projectInfoManager.AddProject
+        | WorkspaceChangeKind.SolutionReloaded ->
+            projectInfoManager.Clear ()
+            args.NewSolution.Projects ?> Seq.iter projectInfoManager.AddProject
+        | WorkspaceChangeKind.SolutionRemoved
+        | WorkspaceChangeKind.SolutionCleared ->
+            projectInfoManager.Clear ()
+        | WorkspaceChangeKind.ProjectAdded ->
+            args.ProjectId |> getProject ?> projectInfoManager.AddProject 
+        | WorkspaceChangeKind.ProjectRemoved ->
+            args.ProjectId |> getProject ?> projectInfoManager.ClearProjectInfo 
+        | WorkspaceChangeKind.ProjectReloaded 
+        | WorkspaceChangeKind.ProjectChanged ->
+            args.ProjectId |> getProject ?> projectInfoManager.UpdateProjectInfo 
+        | WorkspaceChangeKind.DocumentRemoved
+        | WorkspaceChangeKind.DocumentAdded ->
+            args.DocumentId.ProjectId |> getProject ?> projectInfoManager.UpdateProjectInfo 
+        | _ -> ()
+    
 
-    override this.Initialize() =
-        base.Initialize()
-
-        this.Workspace.Options <- this.Workspace.Options.WithChangedOption(Completion.CompletionOptions.BlockForCompletionItems, FSharpConstants.FSharpLanguageName, false)
-        this.Workspace.Options <- this.Workspace.Options.WithChangedOption(Shared.Options.ServiceFeatureOnOffOptions.ClosedFileDiagnostic, FSharpConstants.FSharpLanguageName, Nullable false)
-
-        this.Workspace.DocumentClosed.Add <| fun args ->
-            tryRemoveSingleFileProject args.Document.Project.Id 
-            
-        Events.SolutionEvents.OnAfterCloseSolution.Add <| fun _ ->
-            singleFileProjects.Keys |> Seq.iter tryRemoveSingleFileProject
```

Comment:
That was here, because workspace doesn't trigger DocumentClosed on open files when the solution closes. There was some bug associated with it IIRC.